06_puppy/n0npax/store_test.go

```diff
@@ -0,0 +1,121 @@
+package main
+
+import (
+	"testing"
+
+	tassert "github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/suite"
+)
+
+type storerImpl int
+
+const (
+	syncStorer storerImpl = iota
+	memStorer  storerImpl = iota
+)
+
+var (
+	puppy0 = func() Puppy {
+		return Puppy{
+			Breed:  "Norwegian Forest cat",
+			Colour: "Grey",
+			Value:  "$300",
+		}
+	}
+	puppy1 = func() Puppy {
+		return Puppy{
+			Breed:  "Maine Coon",
+			Colour: "Brown",
+			Value:  "€400",
+		}
+	}
+)
+
+type storerSuite struct {
+	suite.Suite
+	store Storer
+	impl  storerImpl
+}
+
+func (s *storerSuite) SetupTest() {
+	switch s.impl {
+	case syncStorer:
+		s.store = NewSyncStore()
+		puppy := puppy0()
+		s.store.CreatePuppy(&puppy)
+	case memStorer:
+		s.store = NewMemStore()
+		puppy := puppy0()
+		s.store.CreatePuppy(&puppy)
+	default:
+		panic("Unrecognised storer implementation")
+	}
+}
+
+func (s *storerSuite) TestCreatePuppySuccessful() {
+	assert := tassert.New(s.T())
+	newPuppy0, newPuppy1 := puppy0(), puppy1()
+	id0, id1 := s.store.CreatePuppy(&newPuppy0), s.store.CreatePuppy(&newPuppy1)
+	assert.Equal(id0, id1-1, "2nd id should be 1st +1, got", id0, id1)
+}
+
+func (s *storerSuite) TestReadPuppySuccessful() {
+	assert := tassert.New(s.T())
+	expectedPuppy := puppy0()
+	puppy, err := s.store.ReadPuppy(expectedPuppy.ID)
+
+	if assert.NoError(err, "Should be able to read puppy0 from store") {
+		assert.Equal(&expectedPuppy, puppy, "store should return a puppy identical to puppy0")
+	}
+}
+
+func (s *storerSuite) TestReadPuppy_IDDoesNotExist() {
```

Comment:
naming is a bit inconsistent, this test name has an underscore while the others do not

Reply:
You're right, I should avoid underscores in names. Fixed